### ISSUE TITLE
Framework Migration Phase 1

### ISSUE THREAD

**Issue by ahg-g** [kubernetes/kubernetes#83554]
This is an umbrella issue to track Phase 1 of the migration plan as discussed in the [migration doc](https://docs.google.com/document/d/1P1eBaePxuko22B8LLOrPEonoj1awc1dCVBDefuRWUA4/edit#heading=h.u7jfy9wqpd2b).

The main goal of Phase 1 is to change the execution path of predicates and priorities to run through the framework, while retaining the existing execution path until we are confident enough with the new path. We will do this by writing equivalent plugins for all predicates and priorities that call back into the predicates/priorities functions (i.e., no code will be moved or deprecated just yet). 

- [x] Add a translation layer that converts predicate/priority scheduler configuration, whether it comes from the AlgorithmProvider or Policy API, to a Plugin config #82708).

- [x] Refactor predicateMetadata into a collection of sub types (#83193)

- [x] Inject PredicateMetadata and PriorityMetadata into CycleState (#83532) 

- [x]  Make PluginFactoryArgs available to NewDefaultFactory (#83687).

- [x]  Expose ClientSet/SharedInformer in the framework handler (#83523).

- [x] Unblock graduating Component Config to GA so that we can deprecate Policy (#83575)

- [x] Write equivalent Plugins for default predicates. Those plugins should only call the existing predicates logic, NOT move it.

  - [x] Taint and toleration as Filter and Score plugin (#83531).

  - [x] PodFitsResources as Filter plugin (#83630)
  - [x] PodFitsHost as Filter plugin (#83631)
  - [x] PodFitsHostPorts as Filter plugin (#83632)
  - [x] PodMatchNodeSelector as Filter plugin (#83633)
  - [x] NoDiskConflict as Filter plugin (#83629)
  - [x] VolumeBindingChecker as Filter plugin (#83635)
  - [x] CSIMaxVolumeLimitChecker as Filter plugin (#83637)
  - [x] VolumeZoneChecker as Filter plugin (#83634)
  - [x] InterPodAffinityPriority as Filter plugin (#83873)
  - [x] EvenPodsSpread as Filter plugin (#83874)
  - [x] GeneralPred as Filter configuration that combines PodFitsResources, PodFitsHost, PodFitsHostPorts and PodMatchNodeSelector (#83928)
  - [x] CheckNodeUnschedulable as Filter plugin (#84047)

- [x] Write equivalent Plugins for default priorities. Those plugins should only call the existing priorities logic, NOT move it.

  - [x] NodAffinity and Score plugin (#83848)

  - [x] NodePreferAvoidPods as Score plugin (#83770)
  - [x] ImageLocality as Score plugin (#83769)
  - [x] Least/Most/BalancedResourcePriority as Score plugins (#84059)
  - [x] InterPodAffinityPriority as Score plugin
  - [x] EvenPodsSpread as Score plugin
  - [x] SelectorSpread as Score plugin (#84829)
  - [x] Remove EqualPriority priority function, using it in the Policy file should be a no op (#84873)
 

- [x] Equivalent Plugins for custom predicate and priorities. 

  - [x] Disallow duplicate custom plugin configuration (#83472)

  - [x] CheckNodeLabelPresence as Filter plugin (#84187)

  - [x] CheckServiceAffinity as Filter Plugin (#84455)

  - [x] NodeLabelPriority as a Score plugin (#84456)

  - [x] ServiceAntiAffinityPriority as a Score plugin (#84457)

  - [x] RequestedToCapacityRatio as a Score plugin (#84458)

- [x] Allow Cluster Autoscaler to integrate with the Framework, CA will create a Framework rather than a full scheduler (#84834).

**Comment by ahg-g** [comment]
I will create issues for each of the tasks for anyone wants to help and contribute.

**Comment by hprateek43** [comment]
@ahg-g I can work on the tasks. Do let me know a good issue I can start with.

**Comment by damemi** [comment]
I am also available to pick some of these up, will keep an eye out for the sub-issues

**Comment by cofyc** [comment]
hi, @ahg-g 

VolumeBinder initializer needs to access `KubeSchedulerConfiguration.BindTimeoutSeconds` option.

https://github.com/kubernetes/kubernetes/blob/8577711b61cd768df723c6dac333743159393d23/pkg/scheduler/volumebinder/volume_binder.go#L34-L41

Is it possible to expose it in the framework handler too?

**Comment by draveness** [comment]
> Is it possible to expose it in the framework handler too?

I could send another PR for this.

**Comment by cofyc** [comment]
Because this option is only used by volume binder. Alternatively, we can deprecate `KubeSchedulerConfiguration.BindTimeoutSeconds` and use `PluginConfig` to configure volume binder plugin. For backward compatibility, we can inject a `PluginConfig` into `schedulerOptions.frameworkPluginConfig` if `BindTimeoutSeconds` presents and no plugin config for volume binder plugin.
What do you think?

**Comment by cofyc** [comment]
But I'm not sure if it is possible other plugins may need this option in the future.

**Comment by ahg-g** [comment]
> > Is it possible to expose it in the framework handler too?
> 
> I could send another PR for this.

No, this is a plugin configuration parameter, the way to do this is as follows:

1) Add a BindTimeoutSeconds parameter to [ConfigProducerArgs](https://github.com/kubernetes/kubernetes/blob/8577711b61cd768df723c6dac333743159393d23/pkg/scheduler/framework/plugins/default_registry.go#L40)
1) pass this argument as a PluginConfig in the [config producer](https://github.com/kubernetes/kubernetes/blob/8577711b61cd768df723c6dac333743159393d23/pkg/scheduler/framework/plugins/default_registry.go#L61)

**Comment by cofyc** [comment]
Thanks!

**Comment by draveness** [comment]
Can we prioritize #82072 which adds context to the scheduling framework API? The process of converting existing predicates and priorities could cause a lot of rebasing...

**Comment by k82cn** [comment]
@ahg-g , should we highlight the required items in 1.17?

**Comment by lichuqiang** [comment]
@ahg-g I would like to pick of some of the tasks.
So what should I do first? Create issues for related tasks or wait for you to do that?

**Comment by ahg-g** [comment]
Hi @lichuqiang I am creating issues gradually, look for new ones when they become available.

**Comment by ProgramerGu** [comment]
This seems clear，thx

**Comment by ahg-g** [comment]
> @ahg-g , should we highlight the required items in 1.17?

All of them.

**Comment by hprateek43** [comment]
@ahg-g Has someone started to work on `InterPodAffinityPriority as Score plugin` ? If not, can pick it up. Will need some context around it.

**Comment by ahg-g** [comment]
> @ahg-g Has someone started to work on `InterPodAffinityPriority as Score plugin` ? If not, can pick it up. Will need some context around it.

Yeah, this one is tricky, for now we will likely hack it as a NormalizeScore plugin. I am planning to create a couple of other issues, feel free to pick them up.

**Comment by ingvagabund** [comment]
@ahg-g the same question for `EvenPodsSpread as Score plugin`. I can take this one if it's free.

**Comment by ahg-g** [comment]
> @ahg-g the same question for `EvenPodsSpread as Score plugin`. I can take this one if it's free.

@Huang-Wei is working on it.

**Comment by ingvagabund** [comment]
@ahg-g any free task left then? Or, should I wait the same way as @hprateek43 as new issues are presented ? :)

**Comment by josiahbjorgaard** [comment]
Bug triage for 1.17 here. Code freeze is Nov. 18th. Will this issue be closed by then? If not, can we move it out of the milestone?

**Comment by ahg-g** [comment]
we should be able to close it by then, multiple PRs in flight.

**Comment by ahg-g** [comment]
/close

Thanks to all who contributed to this critical task.
